requests-for-comments/2022-11-08-supported-languages.md

```diff
@@ -0,0 +1,169 @@
+# Request for comments: Supported languages at PostHog
+
+Since day 1 here at Posthog we've supported two and half core languages: Python and Javascript/Typescript.
+
+This set of tools has carried us a long way and we should both the cpython runtime and node a huge hats off for bringing us this far.
+
+## Problem statement
+
+We have the opportunity coming up to rebuild or greenfield build out services that are critical parts of our data pipeline. It will be important for these services to be correct, fast, and efficient. Considering this now is a good time to ask: Are we using the correct tools for the job?
+
+So why even bring this up?
+
+Frankly:
+
+- Python is slow.
+- Node is a memory hog.
+- Node tooling is slow.
+- Dependencies can be huge.
+- No guarantees that code is correct.
+
+Can we be successful continuing to use these languages? Of course.
+
+- They are well understood.
+- Our tooling is setup to support these languages
+- We have internal expertise.
+- Generally they are _good enough_.
+
+Why should we be open to other languages?
+
+- We have different people with experience working with different languages.
+- 100% guaranteed static typing is quite beneficial
+- Compiled languages provide more confidence shipped code is correct
+- There are compiled languages that provide significant efficiency wins on CPU, Memory, and performance
+- Using the right tool for the job is typically the right thing to do (if you can ship it)
+
+Candidate Languages:
+
+- [Golang](https://go.dev/) (big surprise)
+- [Rust](https://www.rust-lang.org/) 🦀
+- [OCaml](https://ocaml.org/)
+- [Elixir](https://elixir-lang.org/) (dynamically typed)
+- [Scala](https://www.scala-lang.org/)
+- [Java](<https://en.wikipedia.org/wiki/Java_(programming_language)>)
+
+## Meet the eligible languages
+
+### Golang
+
+The good:
+
+- Extremely easy to learn. Most engineers can learn and be productive in about a day.
+- Built to remove most contentious parts of development
+- Designed to make engineering easier
+- Built for network services and concurrency
+- Great standard lib
+- Super fast compile times
+- Light on memory
+- Used by plenty of organizations big and small
+- We have plenty of Gophers here.
+- A simple binary as a deliverable.
+
+The bad:
+
+- Most people would not say the language sparks joy (I disagree)
+- Not terribly expressive
+- Verbose, but easy to read
+
+### Rust
```

Comment:
I _think_ I would not be wrong in saying I have the most Rust experience here. I LOVE Rust. and I'd love to write it at work. I also have several friends who are fantastic engineers + would consider applying here purely because of Rust. 

But tbh I'd be **super** against us adopting it. The learning curve is huge (prepare to not be comparatively productive for around a quarter). Unless it turns out there's actually more people writing Rust for fun here than I though 😄 

While maturing, the ecosystem is nothing like Python. We'd be spending a lot more time in the weeds. There are fewer "obvious" choices than other languages. EG, which async runtime should we use (if any...)? Why? Do we know enough about how our application does/should work in production to even start making such a choice?

Generally organizations choose Rust because

1. They have explicitly decided that they will be a "Rust company", from the beginning (eg see TrueLayer)
2. They have eked out every last drop of performance from their current stack, and need to go yet further (eg see Discord)